testutil/retry/retry_test.go

```diff
@@ -0,0 +1,47 @@
+package retry
+
+import (
+	"testing"
+	"time"
+)
+
+type failer struct{ calls int }
+
+func (f *failer) FailNow() { f.calls++ }
+
+// delta defines the time band a test run should complete in.
+var delta = 5 * time.Millisecond
+
+func TestRetryer(t *testing.T) {
+	tests := []struct {
+		desc string
+		r    Retryer
+	}{
+		{"counter", &Counter{Count: 3, Wait: 10 * time.Millisecond}},
+		{"timer", &Timer{Timeout: 20 * time.Millisecond, Wait: 10 * time.Millisecond}},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			var n int
+			f := new(failer)
+			start := time.Now()
+			for tt.r.Next(f) {
+				n++
+			}
+			dur := time.Since(start)
+			if got, want := n, 3; got != want {
+				t.Fatalf("got %d retries want %d", got, want)
+			}
+			if got, want := f.calls, 1; got != want {
+				t.Fatalf("got %d FailNow calls want %d", got, want)
+			}
+			// since the first iteration happens immediately
+			// the retryer waits only twice for three iterations.
+			// order of events: (true, wait, true, wait, true, false)
```

Comment:
This should be `(return true, wait and return true, wait and return true, return false)` so that the first operation is executed immediately and you don't have to wait when the exit condition has been met.